packages/core/src/parser/source_map.ts

```diff
@@ -0,0 +1,191 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import _ from 'lodash'
+import { ElemID } from '@salto-io/adapter-api'
+import wu from 'wu'
+import { SourceRange, isSourceRange } from './internal/types'
+
+interface SourceMapEntry {
+  children: Record<string, SourceMapEntry>
+  value: SourceRange[]
+}
+
+const mergeEntries = (src: SourceMapEntry, target: SourceMapEntry): void => {
+  src.value.push(...target.value)
+  // eslint-disable-next-line no-restricted-syntax
+  for (const key in target.children) {
+    if (src.children[key]) {
+      mergeEntries(src.children[key] as SourceMapEntry, target.children[key])
+    } else {
+      src.children[key] = target.children[key]
+    }
+  }
+}
+
+const mountToPath = (
+  data: SourceMapEntry,
+  path: string[],
+  value: SourceMapEntry
+): void => {
+  const [key, ...restOfPath] = path
+  if (!data.children[key]) {
+    data.children[key] = { children: {}, value: [] }
+  }
+  if (_.isEmpty(restOfPath)) {
+    if (data.children[key]) {
+      return mergeEntries(data.children[key], value)
+    }
+    data.children[key] = value
+  }
+  return mountToPath(data.children[key], restOfPath, value)
+}
+
+const setToPath = (
+  data: SourceMapEntry,
+  path: string[],
+  value: SourceRange[]
+): void => {
+  const [key, ...restOfPath] = path
+  if (!data.children[key]) {
+    data.children[key] = { children: {}, value: [] }
+  }
+  if (_.isEmpty(restOfPath)) {
+    data.children[key].value = value
+    return
+  }
+  setToPath(data.children[key], restOfPath, value)
+}
+
+const getFromPath = (
+  data: SourceMapEntry,
+  path: string[],
+  debug = false
+): SourceMapEntry | undefined => {
+  const [key, ...restOfPath] = path
+  if (_.isEmpty(restOfPath)) {
+    return data.children[key]
+  }
+  return data.children[key] ? getFromPath(data.children[key], restOfPath, debug) : undefined
+}
+
+export class SourceMap implements Map<string, SourceRange[]> {
+  [Symbol.toStringTag] = 'SourceMap'
+  private data: SourceMapEntry = { children: {}, value: [] }
+
+  constructor(entries: Iterable<[string, SourceRange[]]> = []) {
+    wu(entries).forEach(([key, value]: [string, SourceRange[]]) => this.set(key, value))
+  }
+
+  private *createGenerator(
+    baseEntry?: SourceMapEntry,
+    prefix: string[] = [],
+  ): Generator<[string, SourceRange[]]> {
+    const data = baseEntry || this.data
+    if (!_.isEmpty(data.value)) yield [prefix.join(ElemID.NAMESPACE_SEPARATOR), data.value]
+    // eslint-disable-next-line no-restricted-syntax
+    for (const key of _.keys(data.children)) {
+      const itr = this.createGenerator(data.children[key], [...prefix, key])
+      yield* itr
+    }
+  }
+
+  [Symbol.iterator](): Generator<[string, SourceRange[]]> {
+    return this.createGenerator()
+  }
+
+  get size(): number { return wu(this.keys()).toArray().length }
```

Comment:
I'm guessing you kept this inefficient implementation because we never actually use this function, right?

anyway, even if we keep the "iterate the map to get the size" thing, we don't have to put everything in memory just to count the number of elements
```suggestion
  get size(): number { return wu.reduce(count => count + 1, 0, this) }
```

Reply:
Yay cool